rama-core/src/matcher/ext.rs

```diff
@@ -0,0 +1,78 @@
+use crate::{
+    matcher::{Extensions, Matcher},
+    Context,
+};
+use std::any::Any;
+
+pub struct ExtensionMatcher<T: Send + Sync> {
+    marker: T,
+}
+
+struct Const<T: Send + Sync>(T);
+struct FnBox(Box<dyn Fn(&dyn Any) -> bool + Send + Sync>);
+
+impl ExtensionMatcher<FnBox> {
+    pub fn with<T: Send + Sync + 'static, F: Fn(&T) -> bool + Send + Sync + 'static>(f: F) -> Self {
+        let wrapped_fn = move |v: &dyn Any| {
+            if let Some(concrete) = v.downcast_ref::<T>() {
+                f(concrete)
+            } else {
+                false
+            }
+        };
+        Self {
+            marker: FnBox(Box::new(wrapped_fn)),
+        }
+    }
+}
+
+impl<T: Send + Sync + Sync + std::cmp::PartialEq> ExtensionMatcher<Const<T>> {
```

Comment:
This can be replaced with the following:

```rust
impl<T> ExtensionMatcher<private::PredicateConst<T>, T>
where
    T: Clone + Eq + Send + Sync + 'static,
{
    /// Create a new [`ExtensionMatcher`] with a const value `T`,
    /// that will be [`Eq`]-checked to find a match.
    pub fn with_const(value: T) -> Self {
        Self {
            predicate: private::PredicateConst(value),
            _marker: PhantomData,
        }
    }
}
```